apicast/src/oauth/apicast_oauth.lua

```diff
@@ -1,26 +1,432 @@
+  ngx.header.content_type = "application/x-www-form-urlencoded"
+  return ngx.redirect(url,"?error=",error.error,"&error_description=",error.error_description,"&state=",state)
+end
+
+function _M.check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+function _M.check_credentials(service, params)
+  local backend = backend_client:new(service)
+
+  local args = {
+      app_id = params.client_id,
+      app_key = params.client_secret,
+      redirect_uri = params.redirect_uri
+    }
+
+  local res = backend:authorize(args)
+
+  return res.status == 200
+end
+
+-- returns a unique string for the client_id. it will be short lived
+local function nonce(client_id)
+  return ts.sha1_digest(tostring(random.bytes(20, true)) .. "#login:" .. client_id)
+end
+
+local function generate_access_token(client_id)
+  local token = ts.sha1_digest(tostring(random.bytes(20, true)) .. client_id)
+
+  return { access_token = token, token_type = "bearer", expires_in = env.get('APICAST_OAUTH_ACCESS_TOKEN_TTL') or 604800 }
+end
+
+local function persist_nonce(service_id, params)
+  -- State value shared between client and gateway
+  local client_state = params.state
+
+  -- State value that will be shared between gateway and authorization server
+  local n = nonce(params.client_id)
+
+  local redis_key = service_id.."#tmp_data:"..n
+  local client_data = {
+    client_id = params.client_id,
+    redirect_uri = params.redirect_uri,
+    plan_id = params.scope,
+    state = client_state,
+    grant = params.response_type
+  }
+
+  local redis = ts.connect_redis()
+
+  if redis then
+    redis:hmset(redis_key, client_data )
+    ts.release_redis(redis)
+  end
+
+  -- Overwrite state to nonce value to share state between gateway and auth server
+  params.state = n
+end
+
+-- Retrieve client data from Redis
+local function retrieve_client_data(service_id, params)
+  local tmp_data = service_id .. "#tmp_data:".. params.state
+  local client_data
+  local redis = ts.connect_redis()
+
+  if redis then
+    local ok, err = redis:hgetall(tmp_data)
+
+    if not ok then
+      ngx.log(0, "no values for tmp_data hash: ".. ts.dump(err))
```

Comment:
right this is probably resource intensive when logging like this...I will look for any other occurrences of logging like this